src/wp-includes/class-wp-block.php

```diff
@@ -333,6 +333,74 @@ private function replace_html( string $block_content, string $attribute_name, $s
 		switch ( $block_type->attributes[ $attribute_name ]['source'] ) {
 			case 'html':
 			case 'rich-text':
+				if ( 'core/image' === $this->name && 'caption' === $attribute_name ) {
+					// Create private anonymous class until the HTML API provides `set_inner_html` method.
+					$bindings_processor = new class( $block_content, WP_HTML_Processor::CONSTRUCTOR_UNLOCK_CODE ) extends WP_HTML_Processor {
+						/**
+						 * Replace the inner content of a figcaption element with the passed content.
+						 *
+						 * THIS IS A TEMPORARY SOLUTION IN CORE NOT TO BE EMULATED.
+						 * IT IS A TEMPORARY SOLUTION THAT JUST WORKS FOR THIS SPECIFIC
+						 * USE CASE UNTIL THE HTML PROCESSOR PROVIDES ITS OWN METHOD.
+						 *
+						 * @since 6.6.0
+						 *
+						 * @param string $new_content New content to insert in the figcaption element.
+						 * @return bool Whether the inner content was properly replaced.
+						 */
+						public function set_content_between_figcaption_balanced_tags( $new_content ) {
+							/*
+							 * THIS IS A STOP-GAP MEASURE NOT TO BE EMULATED.
+							 *
+							 * Check that the processor is paused on an opener tag.
+							 *
+							 */
+							if (
+								WP_HTML_Processor::STATE_MATCHED_TAG !== $this->parser_state ||
+								'FIGCAPTION' !== $this->get_tag() ||
+								$this->is_tag_closer()
+							) {
+								return false;
+							}
+
+							// Set position of the opener tag.
+							$this->set_bookmark( 'opener_tag' );
+
+							/*
+							 * This is a best-effort guess to visit the closer tag and check it exists.
+							 * In the future, this code should rely on the HTML Processor for this kind of operation.
+							 */
+							$tag_name = $this->get_tag();
+							if ( ! $this->next_tag(
+								array(
+									'tag_name'    => $tag_name,
+									'tag_closers' => 'visit',
+								)
+							) || ! $this->is_tag_closer() ) {
+								return false;
+							}
```

Comment:
with the HTML Processor we can put away all mention about "balanced tag" and the like. the system is different - there's no more hacking around the thing that is proper.

a fun fact about the recent push/pop system is that when we reach a point where a node closes, the stack of open elements returns the node that opened it.

```php
public function wait_for_the_real__set_inner_html( $theres_a_reason_not_to_do_this ) {
	/*
	 * It should only report a tag name when matched on a tag, so this
	 * is probably enough and we can skip the parser state check.
	 */
	if ( 'FIGCAPTION' !== $this->get_tag() || $this->is_tag_closer() ) {
		return false;
	}

	// Once this element closes the depth will be one shallower than it is now.
	$depth = $this->get_current_depth();
	while ( $this->next_token() && $this->get_current_depth() >= $depth ) {
		// This is inside the FIGCAPTION element.
	}

	if ( null !== $this->get_last_error() || $this->paused_at_incomplete_token() ) {
		return false;
	}

	$this->set_bookmark( 'here' );
	
	$opening = $this->bookmarks[ $this->current_element->token->bookmark_name ];
	$closing = $this->bookmarks[ '_here' ];
	$start   = $opening->start + $opening->length;

	$this->lexical_updates[] = new WP_HTML_Text_Replacement(
		$start,
		$closing->start - $start,
		wp_kses_post( $theres_a_reason_not_to_do_this )
	};

	return true;
}

$block_reader = $bindings_processor::create_fragment( $block_content );
if ( $block_reader->next_tag( 'figcaption' ) ) {
	$block_reader->wait_for_the_real__set_inner_html( $source_value );
}
return $block_reader->get_updated_html();
```

I've also moved the `wp_kses_post()` call into the processor so it's not exposing the unsafe interface by default. we don't want to rely on people remembering or knowing to do the right thing to avoid the dangerous defaults.

there could be some issues with this; hopefully not, but we've been working to ensure that the depth and the breadcrumbs properly report in the present of virtual nodes. for FIGURE and IMG and FIGCAPTION elements though it's unlikely to find invalid markup. that's way more of a problem for P and TABLE and the like.

just let me know if you find any problems: we can work on it together to make sure it works for you.

Reply:
Thanks a lot for the feedback, @dmsnell  🙂 

I've been testing the suggested logic and it seems to work properly. I just made a commit with it. The only caveat  is that `current_element` is a private property, so I had to make it public. Is this okay or should I look for an alternative method?

Reply:
I'm glad this worked out for you.

regarding `current_element` I'd like to find a way that doesn't involve opening that up to `public`. the most obvious way is simply to create the bookmark on the `FIGCAPTION` opening tag. I want to end up using the trick that the closer references the opener, but maybe that will have to remain inside the class.

for now, can you try setting an `opener` and `closer` bookmark and reference those instead of the `current_element`?

Reply:
> can you try setting an opener and closer bookmark and reference those instead of the current_element?

Okay, I've just reverted the `current_element` change and used an `opener` bookmark. Let me know if that's what you had in mind.